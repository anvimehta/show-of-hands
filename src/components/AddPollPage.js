import React from 'react';
import { connect } from 'react-redux';
import PollForm from './PollForm';
import { startAddPoll } from '../actions/polls';

export class AddPollPage extends React.Component {
  onSubmit = (poll) => {
    this.props.startAddPoll(poll);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Add Poll</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddPoll: (poll) => dispatch(startAddPoll(poll))
});

export default connect(undefined, mapDispatchToProps)(AddPollPage);
