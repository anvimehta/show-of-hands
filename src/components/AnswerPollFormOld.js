import React from 'react';
import { connect } from 'react-redux';
import PollForm from './CreateForm';
import { startEditPoll, startGetuser, startGetPoll } from '../actions/polls';
import getPoll from '../selectors/get-poll';

class AnswerPollForm extends React.Component {
	constructor(props) {
		super(props);
debugger
		this.answerChoices = this.props.poll.choices.slice();
		for (let i = 0; i < this.answerChoices.length; i++) {
			this.answerChoices[i].isSelected = false;
		}
		this.selectedChoice = null;
		this.handleChoiceClick = this.handleChoiceClick.bind(this);
	}

	handleChoiceClick(e) {
		for (let i = 0; i < this.answerChoices.length; i++) {
			if (this.answerChoices[i].text === e.target.value) {
				this.answerChoices[i].isSelected = true;
				this.selectedChoice = this.answerChoices[i];
			} else {
				this.answerChoices[i].isSelected = false;
			}
		}
	}

	handleSubmit() {
		// TODO: Add username to responders array
		const user = startGetUser();
		const poll = this.props.poll;
		if (!poll.responders) {
			poll.responsers = [user.id];
		} else {
			poll.responders = poll.responders.concat([users.id]);
		}
		for (let i = 0; i < poll.choices.length; i++) {
			if (answerChoices[i].isSelected) {
				poll.choices.votes++;
			}
			delete answerChoices[i].isSelected;
		}
		// TODO 
		startEditPoll(poll.id, poll)(() => {
			location.reload();
		});
	}

	render() {
		debugger
		const poll = this.props.poll;

		const choiceButtons = this.answerChoices.map((choice) => (
			<div>
				<input
					type="button"
					value={choice.text}
					className={
						choice.isSelected ? "click-state" : "base-state"
					}
					onClick={e => this.handleChoiceClick(e)}
				/>
			</div>
		));

		return (
			<form onSubmit={this.handleSubmit}>
				<h1>{poll.title}</h1>
				<p>{poll.description}</p>
				{choiceButtons}
				<label>Note: You cannot modify your choice after you submit.</label>
				<input type="submit" value="Submit"/>
			</form>
		);
	}
}

const mapStateToProps = (state) => {
	const poll = getPoll(state.poll);
	return {
	poll
	};
};

const mapDispatchToProps = (dispatch, route) => {
	const pollId = route.location.pathname.match(/\/polls\/(.*)\/answer\/?/)[1];
	return dispatch(startGetPoll({
		id: pollId,
		edit: true
	}));
};

export default (AnswerPollForm);
