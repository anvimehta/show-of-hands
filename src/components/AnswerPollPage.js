import React from 'react';
import { connect } from 'react-redux';
import PollForm from './CreateForm';
import { startAnswerPoll, startGetPoll } from '../actions/polls';
import getPoll from '../selectors/get-poll';

export class AnswerPollPage extends React.Component {
	constructor (props) {
		super(props);
		this.answerPoll = this.answerPoll.bind(this);
		this.handleAnswerChange = this.handleAnswerChange.bind(this);
	}

	answerPoll (poll) {
		const newVoteCount = ++this.props.poll.choices[this.state.selected_answer].votes;
		startAnswerPoll(this.props.poll.id, this.state.selected_answer, this.props.uid, newVoteCount)(() => {
			this.setState({
				poll_answered: true
			});
		});
	}

	handleAnswerChange (e) {
		this.setState({
			selected_answer: e.target.value
		});
	}

	render() {

		if (!this.props.poll || !this.props.poll.id) {
			return null;
		}

		if ((this.state && this.state.poll_answered) || Object(this.props.poll.responses)[this.props.uid]) {
			return (
				<div>
					<div className="page-header">
						<div className="content-container">
							<h1 className="page-header__title">Answer Poll</h1>
						</div>
					</div>
					<div className="content-container">
						<h3>{this.props.poll.title}</h3>
						<p>{this.props.poll.description}</p>
						<div className="answers">
							<p><strong>You have answered the poll.</strong></p>
							<a href={`/dashboard/polls/${this.props.poll.id}`}>
								View the poll
							</a>
						</div>
					</div>
				</div>
			);
		}

		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h1 className="page-header__title">Answer Poll</h1>
					</div>
				</div>
				<div className="content-container">
					<h3>{this.props.poll.title}</h3>
					<p>{this.props.poll.description}</p>
					<div className="answers">
						{
							this.props.poll.choices.map((choice, index) => {
								return <div className="answer-radio" key={index}>
									<label>
										<input
											type="radio"
											name="answer"
											value={index}
											onChange={this.handleAnswerChange}
										/>
										{" "}
										<span>{choice.text}</span>
									</label>
								</div>
							})
						}
					</div><br/>
					<button className="button button--secondary" onClick={this.answerPoll}>Answer Poll</button>
				</div>
			</div>
		);
	}
};

const mapStateToProps = (state) => {
	const poll = getPoll(state.poll);
	const uid = state.auth.uid;
	return { poll, uid };
};

const mapDispatchToProps = (dispatch, route) => {
	const pollId = route.location.pathname.match(/\/polls\/(.*)\/answer\/?/)[1];
	return dispatch(startGetPoll({
		id: pollId
	}));
};

export default connect(mapStateToProps, mapDispatchToProps)(AnswerPollPage);
