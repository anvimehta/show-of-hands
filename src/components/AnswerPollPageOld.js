import React from 'react';
import { connect } from 'react-redux';
import AnswerPollForm from './AnswerPollForm';
import { startGetPoll } from '../actions/polls';

export const AnswerPollPage = (poll) => {
    console.log("poll:", poll);
    return (
        <div>
            <AnswerPollForm poll={poll}/>
        </div>
    );
}

const mapDispatchToProps = (dispatch, route) => {
    const pollId = route.location.pathname.match(/\/polls\/answer\/(.*)/)[1];
    console.log("poll id:", pollId);
    return dispatch(startGetPoll({ id: pollId }));
};

export default connect(undefined, mapDispatchToProps)(AnswerPollPage);
