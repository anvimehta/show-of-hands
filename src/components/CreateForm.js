import React from 'react';
import CATEGORIES, { getCategoryOptions } from '../util/categories.js';

// Component for one choice in the poll
function Choice(props) {
    return (
        <input
            type="text"
            className="choice-title"
            placeholder={"Choice "+props.num}
            value={props.text}
            onChange={(e) => props.onChange(e, props.num-1)}/>
    );
}

class CreateForm extends React.Component {
    constructor(props) {
        super(props);

        this.categories = CATEGORIES;

        this.state = {
            title: '',
            description: '',
            textChoices: ['', ''],
            choices: [{
                text: '',
                votes: 0
            }, {
                text: '',
                votes: 0
            }],
            category: 0,
            newChoiceText: '',
            start_date: new Date(),
            end_date: this.getDate(0),
            public_results: true
        };
        this.updateChoiceValue = this.updateChoiceValue.bind(this);
        this.updateValue = this.updateValue.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }


  componentDidMount () {
    const props = this.props;
    if (props.poll && props.poll.id) {
      this.setState({
            title: props.poll.title,
            description: props.poll.title,
            textChoices: (props.poll.choices || []).map(c => c.text),
            choices: props.poll.choices || [],
            category: props.poll.category,
            newChoiceText: '',
            start_date: props.poll.start_date,
            end_date: props.poll.end_date,
            public_results: props.poll.public_results
        })
    }
  }

    // Methods to update variables when the form is changed by user
    updateValue(e, attr) {
        if (attr === 'title') {
            this.setState({
                title: e.target.value
            });
        } else if (attr === 'description') {
            this.setState({
                description: e.target.value
            });
        } else if (attr === 'newChoiceText') {
            this.setState({
                newChoiceText: e.target.value
            });
        } else if (attr === 'category') {
            this.setState({
                category: e.target.value
            });
        } else if (attr === 'end_date') {
            this.setState({
                end_date: e.target.value
            });
        }
    }
    updateChoiceValue(e, index) {
        // console.log(e.target.value);
        let tc = this.state.textChoices;
        tc[index] = e.target.value;
        let c = this.state.choices;
        c[index].text = e.target.value;
        // console.log(c);
        this.setState({
            textChoices: tc,
            choices: c
        });
    }
    handleNewChoiceClick(e) {
        this.setState({
            textChoices: this.state.textChoices.concat([this.state.newChoiceText]),
            choices: this.state.choices.concat([{
                text: this.state.newChoiceText,
                votes: 0
            }]),
            newChoiceText: ''
        })
    }
    removeChoice(text) {
        let index = 2;
        let tc = this.state.textChoices.slice();
        for (let i = 2; i < tc.length; i++) {
            if (tc[i] === text) {
                index = i;
                break;
            }
        }
        console.log("Remove " + text + ": index " + index);
        console.log(tc);
        tc.splice(index, 1);
        console.log(tc);
        let c = this.state.choices.slice();
        c.splice(index, 1);
        this.setState({
            textChoices: tc,
            choices: c
        });
    }
    updatePublicResults(e) {
        this.setState({
            public_results: !e.target.checked
        })
    }

    // Get current date in YYYY-MM-DD format with an offset of months
    getDate(offset) {
        let date = new Date();
        date.setMonth(date.getMonth() + offset);
        return date.toISOString().substring(0,10);
    }

    // Send form data to backend API
    // TODO: Add user_id
    handleSubmit(e) {
        e.preventDefault()
        this.props.onSubmit({
            title: this.state.title,
            description: this.state.description,
            category: this.state.category,
            choices: this.state.choices,
            start_date: new Date(),
            end_date: this.state.end_date,
            public_results: this.state.public_results,
            responders: [],
            user_id: 0
        });
    }
    render() {
        let i = 1;
        const textChoicesList = this.state.textChoices.map((choice) =>
            <li key={i}>
                <Choice num={i++} text={choice} onChange={this.updateChoiceValue}/>
                { i >= 4 ? (
                    <input type="button" value="-" onClick={() => this.removeChoice(choice)}/>
                ) : null
                }
            </li>
        );
        i = 0;
        const categories = getCategoryOptions()
        return (
            <form onSubmit={this.handleSubmit} >

                {/* Title input */}
                <label>Poll Title</label>
                <input required
                maxLength="40" type="text"
                id="title-input"
                placeholder="Enter a title..."
                value={this.state.title}
                onChange={e => this.updateValue(e, 'title')}/>
                <br/>

                {/* Description input */}
                <label>Poll Description</label>
                <textarea required maxLength="250"
                rows="4" id="description-input"
                placeholder="Enter a short description..."
                value={this.state.description}
                onChange={e => this.updateValue(e, 'description')}></textarea>
                <br/>

                {/* Choice input */}
                <label>Response textChoices</label>
                <ul>{textChoicesList}</ul>
                <br/>

                <input
                    type="text"
                    id="choice-input"
                    value={this.state.newChoiceText}
                    onChange={e => this.updateValue(e, 'newChoiceText')}
                    placeholder="Add a new choice..."/>
                <input type="button" onClick={e => this.handleNewChoiceClick(e)} value="+"></input>
                <br/>

                <label>Poll Category</label>
                <select id="select-category" value={this.state.category}
                onChange={e => this.updateValue(e, 'category')}>
                    {categories}
                </select>
                <br/>

                {/* Lifespan input */}
                <label>This poll will end on:</label>
                <input
                    type="date"
                    value={this.state.end_date}
                    min={this.getDate(0)}
                    max={this.getDate(1)}
                    onChange={e => this.updateValue(e, 'end_date')}/>
                <br/>

                {/* Public results input */}
                <input type="checkbox"
                id="private-results" checked={!this.state.public_results}
                onChange={e => this.updatePublicResults(e)} />
                <label>Make end results private</label>
                <br/>

                <input type="submit" value="Save"/>
            </form>
        )
    }
}

export default CreateForm;
