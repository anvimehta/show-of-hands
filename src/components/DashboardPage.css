body {
    background-color: aqua;
    max-width: 10%;
    border-style: hidden;
    margin: 10%;
}

.create-poll {
    width: 100%;
    display: inline-block;
    border: none;
    padding: 1rem 2rem;
    margin: 0;
    text-decoration: none;
    background: none;
    color: #ffffff;
    font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
    font-variant: small-caps;
    font-size: 2rem;
    cursor: pointer;
    text-align: center;
}

.answer-poll {
    width: 50%;
    display: inline-block;
    border: none;
    padding: 1rem 2rem;
    margin: 0;
    text-decoration: none;
    background: none;
    color: #ffffff;
    font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
    font-variant: small-caps;
    font-size: 2rem;
    cursor: pointer;
    text-align: center;
}

.trending-polls {
    text-align: center;
    color: white;
    font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
}

.bigdiv {
    max-width: 10%;
    border-style: hidden;
    margin: 10%;
}