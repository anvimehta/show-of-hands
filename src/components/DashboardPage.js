import React from 'react';
import { connect } from 'react-redux';
import PollList from './PollList';
import PollListFilters from './PollListFilters';
import PollsSummary from './PollsSummary';
import { listAllPolls } from '../actions/polls';

class DashboardPage extends React.Component {

	constructor (props) {
		super(props);
		this.state = { filters: {} }
		this.rerenderList = this.rerenderList.bind(this)
	}

	rerenderList (filters) {
		/*setState({
			filters
		})*/
		//this.props.refresh()
	}

	render () {
		return <div>
			<PollsSummary />
			<PollListFilters onChange={this.rerenderList} />
			<PollList filters={this.state.filters} />
		</div>
	}
}
 

const mapDispatchToProps = (dispatch) => {
	dispatch(listAllPolls())
	return {
		refresh: () => dispatch(listAllPolls())
	}
};

export default connect(undefined, mapDispatchToProps)(DashboardPage);
