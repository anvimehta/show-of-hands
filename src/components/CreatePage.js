import React from 'react';
import { connect } from 'react-redux';
import CreateForm from './CreateForm.js';
import { listAllPolls, startAddPoll } from '../actions/polls';

class CreatePage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            loading: true
        };
    }

    componentDidMount() {
        listAllPolls()(data => {
            const createdPollsCount = data.polls.reduce((acc, c) => {
                if (c.author === this.props.uid) {
                    return acc + 1
                }
                return acc;
            }, 0);

            const answeredPollsCount = data.polls.reduce((acc, c) => {
                if (c.responses && c.responses[this.props.uid]) {
                    return acc + 1
                }
                return acc;
            }, 0);

            const ratio = answeredPollsCount / createdPollsCount;
            console.log(ratio);

            if (data.polls.length > 5 && ratio < 0.5) {
                this.setState({
                    loading: false,
                    error: "You have to answer some polls. You will be redirected."
                });
                setTimeout(function() { window.location = "/welcome"; }, 2000);
            } else {
                this.setState({ loading: false });
            }
        });
    }

    createPoll(poll) {
        this.props.startAddPoll(poll);
        this.props.history.push('/welcome');
    }

    render() {
        if (this.state.loading) {
            return <div>Loading...</div>;
        }

        if (this.state.error) {
            return <div className="error-alert">{this.state.error}</div>;
        }

        return (
            <div className="create-a-poll">
                <h1>Create a Poll</h1>
                <CreateForm onSubmit={poll => this.createPoll(poll)}/>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddPoll: (poll) => dispatch(startAddPoll(poll))
});

const mapStateToProps = (state) => {
    return { uid: state.auth.uid };
};

export default connect(mapStateToProps, mapDispatchToProps)(CreatePage);
