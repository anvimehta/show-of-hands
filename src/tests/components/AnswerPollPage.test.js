import React from 'react';
import { shallow } from 'enzyme';
import { AnswerPollPage } from '../../components/AnswerPollPage';

test('should correctly render AnswerPollPage', () => {
  const wrapper = shallow(<AnswerPollPage />);
  expect(wrapper).toMatchSnapshot();
});
