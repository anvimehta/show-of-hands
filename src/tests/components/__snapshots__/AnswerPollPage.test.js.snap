// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render AnswerPollPage 1`] = `null`;